warp-tls/Network/Wai/Handler/WarpTLS.hs

```diff
@@ -284,12 +299,13 @@ httpOverTls TLSSettings{..} s bs0 params = do
       , TLS.backendSend  = sendAll s
       , TLS.backendRecv  = recvN
       }
-    conn ctx writeBuf = Connection {
+    conn ctx writeBuf ref = Connection {
         connSendMany         = TLS.sendData ctx . L.fromChunks
       , connSendAll          = sendall
       , connSendFile         = sendfile
       , connClose            = close
-      , connRecv             = recv
+      , connRecv             = recv ref
+      , connRecvBuf          = recvBuf ref
```

Comment:
`Network.Wai.Handler.Warp.Types` says:

```
    -- | The connection receiving function. This returns "" for EOF.
    , connRecv        :: Recv
    -- | The connection receiving function. This tries to fill the buffer.
    --   This returns when the buffer is filled or reaches EOF.
    , connRecvBuf     :: RecvBuf
```

`ref` implements a cache for leftover input data.